src/librbd/api/Group.cc

```diff
@@ -377,6 +394,22 @@ int Group<I>::remove(librados::IoCtx& io_ctx, const char *group_name)
     return r;
   }
 
+  mirror::PromotionState promotion_state;
+  cls::rbd::MirrorGroup mirror_group;
+  C_SaferCond ctx;
+  auto request = mirror::GroupGetInfoRequest<I>::create(
+    io_ctx, "", group_id, &mirror_group, &promotion_state, &ctx);
+  request->send();
+  r = ctx.wait();
+  if (r < 0 && r != -ENOENT && r != -EOPNOTSUPP) {
+    lderr(cct) << "failed to get mirror group info: "
+               << cpp_strerror(r) << dendl;
+    return r;
+  } else if (r == 0 && promotion_state != mirror::PROMOTION_STATE_PRIMARY) {
+    lderr(cct) << "secondary group cannot be removed" << dendl;
```

Comment:
```suggestion
    lderr(cct) << "non-primary group cannot be removed" << dendl;
```
Because we are checking for `!PROMOTION_STATE_PRIMARY` only the group state can be anything other than it (i.e. PROMOTION_STATE_UNKNOWN/PROMOTION_STATE_NON_PRIMARY/PROMOTION_STATE_ORPHAN). We are not quite sure if this is secondary group just by above condition.

Reply:
Should i go with `lderr(cct) << "group is not primary" << dendl;` for all apis?

Reply:
This would be a question to @ajarr.

But if you ask my opinion, I do think it is a good idea to include some meaningful msg about the operation in the log along side the cluster state. For example like:

lderr(cct) << "group is not primary, cannot remove" << dendl;
lderr(cct) << "group is not primary, cannot add image" << dendl;
lderr(cct) << "group is not primary, cannot rename" << dendl;
lderr(cct) << "group is not primary, cannot rename snapshot" << dendl;
lderr(cct) << "group is not primary, cannot perform rollback" << dendl;

and so on.

Reply:
@VinayBhaskar-V as we agreed,

`lderr(cct) << "group is not primary" << dendl;` for all APIs

(or)

`lderr(cct) << "group is not primary, cannot remove" << dendl;`  and similarily for other APIs

(or)

make sure you include `group is not primary` in the error message, and customize it as you see fit.

Pick what makes sense to you